File: modules/core/src/main/scala/com/github/chengpohi/edql/parser/json/JsonCollection.scala
package com.github.chengpohi.edql.parser.json

import com.fasterxml.jackson.core.io.JsonStringEncoder
import org.apache.commons.lang3.{StringEscapeUtils, StringUtils}

import scala.reflect.runtime.universe._

/**
 * eql
 * Created by chengpohi on 2/17/16.
 */
object JsonCollection {

  sealed trait Val extends Any {
    def value: Any

    def apply(i: Int): Val = this.asInstanceOf[Arr].value(i)

    def apply(s: java.lang.String): Val =
      this.asInstanceOf[Obj].value.find(_._1.value == s).get._2

    def toJson: String

    def vars: Seq[Var] = Seq()

    def get(path: String): Option[Val]

    def \\(path: String): Option[Val] = get(path)

    def copy: Val
  }

  case class Str(value: java.lang.String) extends AnyVal with Arith {
    override def toJson: String = "\"" + new String(JsonStringEncoder.getInstance().quoteAsString(value)) + "\""

    def raw: String = value

    override def get(path: String): Option[Val] = None

    override def plus(i: Arith): Arith = i match {
      case n: JsonCollection.Num => JsonCollection.Str(value + n.value.toString)
      case s: JsonCollection.Str => JsonCollection.Str(value + s.value)
      case _ => throw new RuntimeException("not support + type: " + i)
    }

    override def minus(i: Arith): Arith = throw new RuntimeException("not support + type: " + i)

    override def multiply(i: Arith): Arith = throw new RuntimeException("not support + type: " + i)

    override def div(i: Arith): Arith = throw new RuntimeException("not support + type: " + i)

    override def copy: Val = Str(value)
  }

  abstract class Dynamic extends Val

  case class Var(value: java.lang.String) extends Dynamic {
    var realValue: Option[JsonCollection.Val] = None

    override def toJson: String = realValue.map(_.toJson).getOrElse("")

    override def get(path: String): Option[Val] = None

    override def vars: Seq[Var] = Seq(this)

    override def copy: Val = {
      val va = Var(value)
      va.realValue = realValue.map(_.copy)
      va
    }
  }

  case class Fun(value: (String, Seq[Val])) extends Dynamic {
    var realValue: Option[JsonCollection.Val] = None

    override def toJson: String = realValue.map(_.toJson).getOrElse("")

    override def get(path: String): Option[Val] = None

    override def copy: Val = {
      val f = Fun((value._1, value._2.map(i => i.copy)))
      f.realValue = realValue.map(_.copy)
      f
    }
  }

  case class ArithTree(value: (JsonCollection.Val, Option[String], Option[JsonCollection.Val])) extends Dynamic {
    var realValue: Option[JsonCollection.Arith] = None

    override def toJson: String = realValue.map(_.toJson).getOrElse(value._1.toJson)

    override def get(path: String): Option[Val] = None

    override def copy: Val = {
      val tree = ArithTree((value._1.copy, value._2, value._3.map(_.copy)))
      tree.realValue = realValue.map(_.copy.asInstanceOf[Arith])
      tree
    }
  }

  case class Obj(value: (Val, Val)*) extends AnyVal with Val {
    override def toJson: String = {
      val valueJson = value.map {
        case (n, v) => {
          val j = v.toJson
          n.toJson + ":" + j
        }
      }.filter(i => StringUtils.isNotBlank(i)).mkString(",")
      "{" + valueJson + "}"
    }

    override def get(path: String): Option[Val] =
      value.find(p => p._1.value == path).map(_._2)

    override def vars: Seq[Var] = this.value.flatMap(i => i._1.vars ++ i._2.vars)

    def remove(v: String): Obj = {
      val nvs = value.filter(i => i._1.asInstanceOf[Str].value != v)
      Obj(nvs: _*)
    }

    def add(k: String, v: Val): Obj = {
      val nvs = value :+ (Str(k), v)
      Obj(nvs: _*)
    }

    override def copy: Val = Obj(value.map(i => (i._1.copy, i._2.copy)): _*)
  }

  case class Arr(value: Val*) extends AnyVal with Val {
    override def toJson: String =
      "[" + value.map(i => i.toJson).mkString(",") + "]"

    override def get(path: String): Option[Val] = None

    override def vars: Seq[Var] = this.value.flatMap(_.vars)

    override def copy: Val = Arr(value.map(_.copy): _*)
  }

  case class Tuple(value: Val*) extends AnyVal with Val {
    override def toJson: String =
      "(" + value.map(i => i.toJson).mkString(",") + ")"

    override def get(path: String): Option[Val] = None

    override def vars: Seq[Var] = value.flatMap(_.vars)

    override def copy: Val = Tuple(value.map(_.copy): _*)
  }

  trait Arith extends Any with Val {
    def plus(i: Arith): Arith

    def minus(i: Arith): Arith

    def multiply(i: Arith): Arith

    def div(i: Arith): Arith
  }

  case class Num(value: Number) extends AnyVal with Arith {
    override def toJson: String = java.math.BigDecimal.valueOf(value.doubleValue()).stripTrailingZeros().toPlainString;


    override def get(path: String): Option[Val] = None

    override def plus(i: Arith): Arith = {
      i match {
        case n: JsonCollection.Num => JsonCollection.Num(addNumbers(value, n.value))
        case s: JsonCollection.Str => JsonCollection.Str(value + s.value)
        case _ => throw new RuntimeException("not support + type: " + i)
      }
    }

    override def minus(i: Arith): Arith = {
      i match {
        case n: JsonCollection.Num => JsonCollection.Num(minusNumbers(value, n.value))
        case _ => throw new RuntimeException("not support + type: " + i)
      }
    }

    override def multiply(i: Arith): Arith = {
      i match {
        case n: JsonCollection.Num => JsonCollection.Num(multiplyNumbers(value, n.value))
        case _ => throw new RuntimeException("not support + type: " + i)
      }
    }

    override def div(i: Arith): Arith = {
      i match {
        case n: JsonCollection.Num => JsonCollection.Num(divNumbers(value, n.value))
        case _ => throw new RuntimeException("not support + type: " + i)
      }
    }

    override def copy: Val = Num(value)
  }

  def addNumbers(a: Number, b: Number): Number = {
    if (a.isInstanceOf[Double] || b.isInstanceOf[Double]) a.doubleValue + b.doubleValue
    else if (a.isInstanceOf[Float] || b.isInstanceOf[Float]) a.floatValue + b.floatValue
    else if (a.isInstanceOf[Long] || b.isInstanceOf[Long]) a.longValue + b.longValue
    else a.intValue + b.intValue
  }

  def minusNumbers(a: Number, b: Number): Number = {
    if (a.isInstanceOf[Double] || b.isInstanceOf[Double]) a.doubleValue - b.doubleValue
    else if (a.isInstanceOf[Float] || b.isInstanceOf[Float]) a.floatValue - b.floatValue
    else if (a.isInstanceOf[Long] || b.isInstanceOf[Long]) a.longValue - b.longValue
    else a.intValue - b.intValue
  }

  case object False extends Val {
    def value = false

    override def toJson: String = value.toString

    override def get(path: String): Option[Val] = None

    override def copy: Val = False
  }

  case object True extends Val {
    def value: Boolean = true

    override def toJson: String = value.toString

    override def get(path: String): Option[Val] = None

    override def copy: Val = True
  }

  case object Null extends Val {
    def value: Option[Nothing] = None

    override def toJson: String = value.map(_.toString).orNull

    override def get(path: String): Option[Val] = None

    override def copy: Val = Null
  }


  def multiplyNumbers(a: Number, b: Number): Number = {
    if (a.isInstanceOf[Double] || b.isInstanceOf[Double]) a.doubleValue * b.doubleValue
    else if (a.isInstanceOf[Float] || b.isInstanceOf[Float]) a.floatValue * b.floatValue
    else if (a.isInstanceOf[Long] || b.isInstanceOf[Long]) a.longValue * b.longValue
    else a.intValue * b.intValue
  }

  def divNumbers(a: Number, b: Number): Number = {
    if (a.isInstanceOf[Double] || b.isInstanceOf[Double]) a.doubleValue / b.doubleValue
    else if (a.isInstanceOf[Float] || b.isInstanceOf[Float]) a.floatValue / b.floatValue
    else if (a.isInstanceOf[Long] || b.isInstanceOf[Long]) a.longValue / b.longValue
    else a.intValue / b.intValue
  }

  case object Comment extends Val {
    def value: Option[Nothing] = None

    override def toJson: String = value.toString

    override def get(path: String): Option[Val] = None

    override def copy: Val = Comment
  }

  implicit class JsonConverter(value: Val) {
    private[JsonCollection] def extract(tag: Type): Any = {
      if (tag <:< typeOf[Map[_, _]]) {
        val subType2: Type = tag.typeArgs(1)
        value
          .asInstanceOf[Obj]
          .value
          .toList
          .map(i => (i._1, i._2.extract(subType2)))
          .toMap
      } else if (tag <:< typeOf[List[(_, _)]]) {
        val subType: Type = tag.typeArgs.head
        value
          .asInstanceOf[Obj]
          .value
          .toList
          .map(i => (i._1, i._2.extract(subType.typeArgs(1))))
      } else if (tag <:< typeOf[List[_]]) {
        val subType: Type = tag.typeArgs.head
        value
          .asInstanceOf[Arr]
          .value
          .toList
          .map(i => {
            i.extract(subType)
          })
      } else if (tag <:< typeOf[(_, _)]) {
        val subType = tag.typeArgs
        val (tp1, tp2) = (subType.head, subType(1))
        val vals = value.asInstanceOf[Arr].value.toList
        (vals.head.extract(tp1), vals(1).extract(tp2))
      } else if (tag =:= typeOf[Int]) {
        value.asInstanceOf[Num].value
      } else {
        value.value
      }
    }

    def extract[T](implicit tag: TypeTag[T]): T =
      extract(tag.tpe).asInstanceOf[T]
  }

}

File: src/main/java/com/github/chengpohi/edql/parser/json/JsonCollection.kt
package com.github.chengpohi.edql.parser.json

import com.fasterxml.jackson.core.io.JsonStringEncoder
import org.apache.commons.lang3.StringUtils
import java.math.BigDecimal

object JsonCollection {
    sealed interface Val {
        val value: Any?

        fun apply(i: Int): Val = (this as Arr).value[i]

        fun apply(s: String): Val =
            (this as Obj).value.find { it.first.value == s }?.second
                ?: throw NoSuchElementException("Key not found: $s")

        fun toJson(): String

        fun vars(): List<Var> = emptyList()

        fun funs(): List<Fun> = emptyList()

        fun get(path: String): Val?

        fun copy(): Val
    }

    data class Str(override val value: String) : Val, Arith {
        override fun toJson(): String {
            return "\"${String(JsonStringEncoder.getInstance().quoteAsString(value))}\""
        }

        fun raw(): String = value

        override fun get(path: String): Val? = null

        override fun plus(i: Arith): Arith = when (i) {
            is Num -> Str(value + i.value.toString())
            is Str -> Str(value + i.value)
            else -> throw RuntimeException("not support + type: $i")
        }

        override fun minus(i: Arith): Arith = throw RuntimeException("not support - type: $i")

        override fun multiply(i: Arith): Arith = throw RuntimeException("not support * type: $i")

        override fun div(i: Arith): Arith = throw RuntimeException("not support / type: $i")

        override fun copy(): Val = this
    }

    abstract class Dynamic : Val {
        abstract fun clean()

        var realValue: Val? = null
    }

    data class Var(override val value: String) : Dynamic() {
        override fun toJson(): String = realValue?.toJson() ?: ""

        override fun get(path: String): Val? = null

        override fun vars(): List<Var> = listOf(this)

        override fun copy(): Val {
            val va = Var(value)
            va.realValue = realValue?.copy()
            return va.realValue ?: this
        }

        override fun clean() {
            realValue = null
        }
    }

    data class Fun(override val value: Pair<String, List<Val>>) : Dynamic() {

        override fun toJson(): String = realValue?.toJson() ?: ""

        override fun get(path: String): Val? = null

        override fun funs(): List<Fun> = listOf(this)

        override fun copy(): Val {
            val f = Fun(value.first to value.second.map { it.copy() })
            return realValue?.copy() ?: this
        }

        override fun clean() {
            value.second.filterIsInstance<Dynamic>().forEach { it.clean() }
            realValue = null
        }
    }

    data class ArithTree(
        var a: Val,
        var op: String?,
        var b: Val?,
        var order: Int? = null
    ) : Dynamic() {
        override val value: Triple<Val, String?, Val?> = Triple(a, op, b)

        override fun toJson(): String = realValue?.toJson() ?: a.toJson()

        override fun get(path: String): Val? = null

        override fun copy(): Val {
            val tree = ArithTree(a.copy(), op, b?.copy(), null)
            return realValue?.copy() as? Arith ?: tree
        }

        override fun clean() {
            when (a) {
                is Dynamic -> (a as Dynamic).clean()
                else -> {}
            }
            a.vars().forEach { it.clean() }
            b?.let { o ->
                if (o is Dynamic) {
                    o.clean()
                }
                o.vars().forEach { it.clean() }
            }
            realValue = null
        }
    }

    data class Obj(override val value: List<Pair<Val, Val>>) : Val, Arith {
        constructor(vararg pairs: Pair<Val, Val>) : this(pairs.toList())

        override fun toJson(): String {
            val valueJson = value.map { (n, v) ->
                val j = v.toJson()
                "${n.toJson()}:$j"
            }.filter { StringUtils.isNotBlank(it) }.joinToString(",")
            return "{$valueJson}"
        }

        override fun get(path: String): Val? =
            value.find { it.first.value == path }?.second

        override fun vars(): List<Var> = value.flatMap { (k, v) -> k.vars() + v.vars() }

        override fun funs(): List<Fun> = value.flatMap { (k, v) -> k.funs() + v.funs() }

        fun remove(v: String): Obj {
            val nvs = value.filter { it.first !is Str || (it.first as Str).value != v }
            return Obj(nvs)
        }

        fun add(k: String, v: Val): Obj {
            val nvs = value + (Str(k) to v)
            return Obj(nvs)
        }

        override fun copy(): Val = Obj(value.map { (k, v) -> k.copy() to v.copy() })

        override fun plus(i: Arith): Arith = when (i) {
            is Obj -> Obj(value + i.value)
            else -> throw RuntimeException("not support + type: $i")
        }

        override fun minus(i: Arith): Arith = throw RuntimeException("not support - type: $i")

        override fun multiply(i: Arith): Arith = throw RuntimeException("not support * type: $i")

        override fun div(i: Arith): Arith = throw RuntimeException("not support / type: $i")
    }

    data class Arr(override val value: List<Val>) : Val {
        constructor(vararg values: Val) : this(values.toList())
        override fun toJson(): String =
            "[${value.joinToString(",") { it.toJson() }}]"

        override fun get(path: String): Val? = null

        override fun vars(): List<Var> = value.flatMap { it.vars() }

        override fun funs(): List<Fun> = value.flatMap { it.funs() }

        override fun copy(): Val = Arr(value.map { it.copy() })
    }

    data class Tuple(override val value: List<Val>) : Val {
        constructor(vararg values: Val) : this(values.toList())

        override fun toJson(): String =
            "(${value.joinToString(",") { it.toJson() }})"

        override fun get(path: String): Val? = null

        override fun vars(): List<Var> = value.flatMap { it.vars() }

        override fun funs(): List<Fun> = value.flatMap { it.funs() }

        override fun copy(): Val = Tuple(value.map { it.copy() })
    }

    interface Arith : Val {
        fun plus(i: Arith): Arith
        fun minus(i: Arith): Arith
        fun multiply(i: Arith): Arith
        fun div(i: Arith): Arith
    }

    data class Num(override val value: Number) : Val, Arith {
        override fun toJson(): String {
            return when (value) {
                is BigDecimal -> value.toString()
                else -> BigDecimal.valueOf(value.toDouble()).stripTrailingZeros().toPlainString()
            }
        }

        override fun get(path: String): Val? = null

        override fun plus(i: Arith): Arith = when (i) {
            is Num -> Num(addNumbers(value, i.value))
            is Str -> Str(value.toString() + i.value)
            else -> throw RuntimeException("not support + type: $i")
        }

        override fun minus(i: Arith): Arith = when (i) {
            is Num -> Num(minusNumbers(value, i.value))
            else -> throw RuntimeException("not support - type: $i")
        }

        override fun multiply(i: Arith): Arith = when (i) {
            is Num -> Num(multiplyNumbers(value, i.value))
            else -> throw RuntimeException("not support * type: $i")
        }

        override fun div(i: Arith): Arith = when (i) {
            is Num -> Num(divNumbers(value, i.value))
            else -> throw RuntimeException("not support / type: $i")
        }

        override fun copy(): Val = this
    }

    fun addNumbers(a: Number, b: Number): Number {
        return when {
            a is Double || b is Double -> a.toDouble() + b.toDouble()
            a is Float || b is Float -> a.toFloat() + b.toFloat()
            a is Long || b is Long -> a.toLong() + b.toLong()
            else -> a.toInt() + b.toInt()
        }
    }

    fun minusNumbers(a: Number, b: Number): Number {
        return when {
            a is Double || b is Double -> a.toDouble() - b.toDouble()
            a is Float || b is Float -> a.toFloat() - b.toFloat()
            a is Long || b is Long -> a.toLong() - b.toLong()
            else -> a.toInt() - b.toInt()
        }
    }

    object False : Val {
        override val value: Boolean = false

        override fun toJson(): String = value.toString()

        override fun get(path: String): Val? = null

        override fun copy(): Val = this
    }

    object True : Val {
        override val value: Boolean = true

        override fun toJson(): String = value.toString()

        override fun get(path: String): Val? = null

        override fun copy(): Val = this
    }

    object Null : Val {
        override val value: Nothing? = null

        override fun toJson(): String = value?.toString() ?: "null"

        override fun get(path: String): Val? = null

        override fun copy(): Val = this
    }

    fun multiplyNumbers(a: Number, b: Number): Number {
        return when {
            a is Double || b is Double -> a.toDouble() * b.toDouble()
            a is Float || b is Float -> a.toFloat() * b.toFloat()
            a is Long || b is Long -> a.toLong() * b.toLong()
            else -> a.toInt() * b.toInt()
        }
    }

    fun divNumbers(a: Number, b: Number): Number {
        return when {
            a is Double || b is Double -> a.toDouble() / b.toDouble()
            a is Float || b is Float -> a.toFloat() / b.toFloat()
            a is Long || b is Long -> a.toLong() / b.toLong()
            else -> a.toInt() / b.toInt()
        }
    }

    object Comment : Val {
        override val value: Nothing? = null

        override fun toJson(): String = value.toString()

        override fun get(path: String): Val? = null

        override fun copy(): Val = this
    }
}